### ISSUE TITLE
DLL load failure when importing after PyTorch

### ISSUE THREAD

**Issue by peterjc123** [scipy/scipy#12400]
My issue is about DLL load failure on Windows. We are using Anaconda. It is causing test jobs to fail at PyTorch: https://github.com/pytorch/pytorch/issues/40366. Currently, we are downgrading it to 1.4.1.

#### Reproducing code example:

```cmd
pip install --pre scipy torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
python -c "import torch; from scipy import stats"
```

#### Error message:

```pytb
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Jenkins\Miniconda3\lib\site-packages\scipy\__init__.py", line 104, in <module>
    from . import _distributor_init
  File "C:\Jenkins\Miniconda3\lib\site-packages\scipy\_distributor_init.py", line 61, in <module>
    WinDLL(os.path.abspath(filename))
  File "C:\Jenkins\Miniconda3\lib\ctypes\__init__.py", line 348, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found
```

#### Scipy/Numpy/Python version information:
```
1.5.0 1.18.1 sys.version_info(major=3, minor=6, micro=7, releaselevel='final', serial=0)
```

**Comment by andyfaff** [comment]
There's trace on the pytorch issue, and it's failing to load libopenblas.

**Comment by tylerjereddy** [comment]
I thought we just dealt with another round of `_distributor_init.py` issues recently. Is this reproducible in a clean virtualenv instead of using `conda`?

**Comment by tylerjereddy** [comment]
There are a bunch of `gfortran` loading complaints there too, I think.

**Comment by imatiach-msft** [comment]
+1 seeing this issue as well in our builds after latest scipy release
Adding print in my env:

```
            print(self._name)
            self._handle = _dlopen(self._name, mode)
```

shows:
```
kernel32
None
rpcrt4
kernel32
user32
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\sklearn\.libs\vcomp140.dll
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\scipy\.libs\libansari.R6EA3HQP5KZ6TAXU4Y4ZVTRPT7UVA53Z.gfortran-win_amd64.dll
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\scipy\.libs\libbanded5x.IK367JJHLT65TXN7P5ACJU7BJJIBU2SP.gfortran-win_amd64.dll
```
And then the stack trace:

```
ImportError while loading conftest 'C:\azuremlcli\tests\scenarios\explain_model\conftest.py'.
..\conftest.py:9: in <module>
    from explain_model.common_utils import create_iris_data, create_boston_data, create_simple_titanic_data, \
..\common_utils.py:9: in <module>
    from sklearn import svm, ensemble, linear_model
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\sklearn\__init__.py:80: in <module>
    from .base import clone
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\sklearn\base.py:21: in <module>
    from .utils import _IS_32BIT
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\sklearn\utils\__init__.py:20: in <module>
    from scipy.sparse import issparse
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\scipy\__init__.py:104: in <module>
    from . import _distributor_init
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\site-packages\scipy\_distributor_init.py:61: in <module>
    WinDLL(os.path.abspath(filename))
C:\Users\ilmat\AppData\Local\Continuum\Miniconda3\envs\test3\lib\ctypes\__init__.py:349: in __init__
    self._handle = _dlopen(self._name, mode)
E   OSError: [WinError 126] The specified module could not be found
```

**Comment by tylerjereddy** [comment]
@imatiach-msft How are you installing SciPy? `conda` or `pip`?

@jjhelmus @rgommers Maybe something isn't playing nicely between the new `_distributor_init.py` and conda environments?

If this is a SciPy wheel installed in a conda env, are we responsible for making sure that our `_distributor_init.py` supports that scenario?

If this is a conda distribution of SciPy, then the usage of our `_distributor_init.py` may not necessarily be recommended anyway.

**Comment by peterjc123** [comment]
@tylerjereddy If that is the case, maybe you could add an if-condition check like this.
```py
if os.name == 'nt' and not os.path.exists(sys.prefix, 'conda-meta'):
```

**Comment by imatiach-msft** [comment]
@tylerjereddy  scipy is installed via pip, it seems to be some bad interaction with pytorch

**Comment by ilayn** [comment]
Pip and conda installs often don't play nice together. Could you please check if the problem is still there if you conda install SciPy?

**Comment by rgommers** [comment]
> @tylerjereddy scipy is installed via pip, it seems to be some bad interaction with pytorch. pytorch is installed via conda in my build.

That's indeed not the best idea, but it used to work - I suspect if we fix this issue for both-pip-installed `pytorch` and `scipy`, this will work again. But yeah ..... don't do this if you can avoid it.

> Could you please check if the problem is still there if you conda install SciPy?

Since our conda builds don't have `_distributor_init.py`, that won't show this failure.

> If this is a SciPy wheel installed in a conda env, are we responsible for making sure that our _distributor_init.py supports that scenario?
> 
> If this is a conda distribution of SciPy, then the usage of our _distributor_init.py may not necessarily be recommended anyway.

Yes that should work, and we will need the same `_distributor_init.py` - that mechanism is related to loaded libraries shipped in our wheels, not with conda Python vs. other Python.

**Comment by rgommers** [comment]
I don't immediately see how the change in https://github.com/MacPython/scipy-wheels/pull/70 would be responsible, it just adds a `chdir` and the `abspath` handling looks fine. However it is the most obvious candidate, and it's easy to test if one can reproduce the issue (I don't have a Windows box at hand): in the `pip`-installed `scipy`, edit `_distributor_init.py` to revert the changes from https://github.com/MacPython/scipy-wheels/pull/70, then do `python -c "import torch; from scipy import stats"` and see if that works or not.

**Comment by rgommers** [comment]
Oh wait, that version comparison change in that scipy-wheels PR made a much larger change:
```
$ diff scipy-1.4.1-win64/_distributor_init.py scipy-1.5.0-win64/_distributor_init.py 
10a11,63
> 
> import os
> 
> # on Windows SciPy loads important DLLs
> # and the code below aims to alleviate issues with DLL
> # path resolution portability with an absolute path DLL load
> if os.name == 'nt':
>     from ctypes import WinDLL
>     import glob
>     # convention for storing / loading the DLL from
>     # scipy/.libs/, if present
>     libs_path = os.path.abspath(os.path.join(os.path.dirname(__file__),
>                                              '.libs'))
>     if os.path.isdir(libs_path):
>         # for Python >= 3.8, DLL resolution ignores the PATH variable
>         # and the current working directory; see release notes:
>         # https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
>         # Only the system paths, the directory containing the DLL, and 
>         # directories added with add_dll_directory() are searched for 
>         # load-time dependencies with Python >= 3.8
> 
>         # this module was originally added to support DLL resolution in
>         # Python 3.8 because of the changes described above--providing the
>         # absolute paths to the DLLs allowed for proper resolution/loading
> 
>         # however, we also started to receive reports of problems with DLL
>         # resolution with Python 3.7 that were sometimes alleviated with
>         # inclusion of the _distributor_init.py module; see SciPy main 
>         # repo gh-11826
> 
>         # we noticed in scipy-wheels repo gh-70 that inclusion of
>         # _distributor_init.py in 32-bit wheels for Python 3.7 resulted
>         # in failures in DLL resolution (64-bit 3.7 did not)
>         # as a result, we decided to combine both the old (working directory)
>         # and new (absolute path to DLL location) DLL resolution mechanisms
>         # to improve the chances of resolving DLLs across a wider range of
>         # Python versions
> 
>         # we did not experiment with manipulating the PATH environment variable
>         # to include libs_path; it is not immediately clear if this would have
>         # robustness or security advantages over changing working directories
>         # as done below
> 
>         # we should remove the working directory shims when our minimum supported
>         # Python version is 3.8 and trust the improvements to secure DLL loading
>         # in the standard lib for Python >= 3.8
>         try:
>             owd = os.getcwd()
>             os.chdir(libs_path)
>             for filename in glob.glob(os.path.join(libs_path, '*dll')):
>                 WinDLL(os.path.abspath(filename))
>         finally:
>             os.chdir(owd)
```

The above is from the win64 wheels downloaded from PyPI. It may be the case that we're now doing `WinDLL` calls where we weren't doing any before (on Python 3.6). PyTorch CI only has Python 3.6, so can't tell if there was an issue here with 3.8, which is the only version we did these `WinDLL` calls for previously.

**Comment by peterjc123** [comment]
@rgommers Yes, we have a stacktrace in the pytorch issue. It shows the root cause of the failure. When it tries to load `scipy\.libs\libbanded5x.IK367JJHLT65TXN7P5ACJU7BJJIBU2SP.gfortran-win_amd64.dll`, the os finds out that it relies on `libopenblas.3HBPCJB5BPQGKWVZAVEBXNNJ2Q2G3TUP.gfortran-win_amd64.dll` and thus tries to load that DLL, but failed. Looks like using `chdir` here fails to add a search path.

**Comment by peterjc123** [comment]
As documented in https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order, there are several cases that the current directory is not considered as a DLL search path.
1. When `SetDllDirectory` is used previously (this seems to be the case for Anaconda `ctypes.WinDLL`)
2. `LoadLibraryEx` with LOAD_LIBRARY_SEARCH flags (but this case is fine because we may use LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR to add the directory of the original lib into search paths)

**Comment by ilayn** [comment]
I'm shooting in the dark but is this somehow still related to https://github.com/numpy/numpy/pull/14974 ?

**Comment by rgommers** [comment]
This is the "old way" that's in `scipy/__config__.py`, that works for older Windows versions and is still there:
```
extra_dll_dir = os.path.join(os.path.dirname(__file__), '.libs')

if sys.platform == 'win32' and os.path.isdir(extra_dll_dir):
    os.environ.setdefault('PATH', '')
    os.environ['PATH'] += os.pathsep + extra_dll_dir
```

Ah I think I see what the issue is: the above code snippet isn't yet run because `__config__` is only imported in `__init__.py` *after* `from . import _distributor_init.py`. Swapping that around should resolve this issue I think.

> I'm shooting in the dark but is this somehow still related to numpy/numpy#14974 ?

Yes it is related to that whole mess.

> 1. When `SetDllDirectory` is used previously (this seems to be the case for this issue)

I guess that's why it shows up only after `import torch`? Makes sense.

**Comment by peterjc123** [comment]
> I guess that's why it shows up only after `import torch`? Makes sense.

But I didn't find where `SetDllDirectory` is called in PyTorch. We only called `AddDllDirectory` and `LoadLibrarayEx` in `__init__.py`.

**Comment by rgommers** [comment]
@peterjc123 there's a bunch of dependencies though, one of which is likely doing this (e.g. MKL).

**Comment by rgommers** [comment]
If someone can confirm that adding `from . import __config__` to `__init__.py` right above `from . import _distributor_init`, that would be super useful.

**Comment by pv** [comment]
The `from . import __config__` could also be added to `_distributor_init.py`.

Also, the numpy-distutils submodule in scipy-wheels maybe needs to be bumped to recent numpy, so that the `__config__.py` snippet is updated, to use `os.add_dll_directory` on python3.8 instead of PATH manipulations.

**Comment by tylerjereddy** [comment]
When I try running the original reproducer in a fresh Python 3.8 conda env on Windows 10 (native powershell) I don't see a valid install working:

```
pip install scipy torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
```

```
  ERROR: Command errored out with exit status 1:
   command: 'C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\treddy\\AppData\\Local\\Temp\\pip-install-3uzu7sxi\\torch\\setup.py'"'"'; __file__='"'"'C:\\Users\\treddy\\AppData\\Local\\Temp\\pip-install-3uzu7sxi\\torch\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\treddy\AppData\Local\Temp\pip-wheel-pjfya69k'
       cwd: C:\Users\treddy\AppData\Local\Temp\pip-install-3uzu7sxi\torch\
  Complete output (30 lines):
  running bdist_wheel
  running build
  running build_deps
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "C:\Users\treddy\AppData\Local\Temp\pip-install-3uzu7sxi\torch\setup.py", line 225, in <module>
      setup(name="torch", version="0.1.2.post2",
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\site-packages\setuptools\__init__.py", line 161, in setup
      return distutils.core.setup(**attrs)
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\core.py", line 148, in setup
      dist.run_commands()
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\site-packages\wheel\bdist_wheel.py", line 223, in run
      self.run_command('build')
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\command\build.py", line 135, in run
      self.run_command(cmd_name)
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\Users\treddy\windows_dev_work\conda_install\envs\test_env\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "C:\Users\treddy\AppData\Local\Temp\pip-install-3uzu7sxi\torch\setup.py", line 51, in run
      from tools.nnwrap import generate_wrappers as generate_nn_wrappers
  ModuleNotFoundError: No module named 'tools.nnwrap'
  ----------------------------------------
  ERROR: Failed building wheel for torch
```

**Comment by tylerjereddy** [comment]
If I use the work-around floating around the internet (...) `pip install scipy torch==1.4.0+cpu torchvision==0.5.0+cpu -f https://download.pytorch.org/whl/torch_stable.html`, I don't see the DLL load issue reported above in Python 3.8 conda env with:

```python
In [1]: import torch
In [2]: from scipy import stats
```

I'll try with Python 3.6 next..

**Comment by tylerjereddy** [comment]
I can reproduce in a fresh conda 3.6 env, but not 3.8.

**Comment by peterjc123** [comment]
@tylerjereddy Sorry for posting the wrong command, actually it should be `pip install --pre scipy torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html`. I've updated the one in the description as well.

**Comment by peterjc123** [comment]
> I can reproduce in a fresh conda 3.6 env, but not 3.8.

Yes, observed the same result. It seems the load library semantics are different in python 3.8.

**Comment by rgommers** [comment]
That makes sense, because we introduced the `WinDLL` calls in response to a change in Python 3.8 in Windows store, that was later backported to 3.7. Python 3.6 never got that change, but we did lose setting the path to include `.libs` for 3.6 - hence the issue.

**Comment by tylerjereddy** [comment]
I've tried to take some of the suggestions proposed here and use them in: https://github.com/MacPython/scipy-wheels/pull/88

That said, I actually do still have problems consistently reproducing the issue locally on Windows 10 now that I look more closely. Mixing conda and pip can cause a variety of issues so it is actually pretty easy to pollute the "testing" I think.

A simple regression guard for that wheels PR would be nice (i.e., a really clean reproducer, even if artificial), though I suppose our natural progression beyond Python 3.6 will at least prevent this from becoming an unusually large maintenance risk/burden in the long term.

**Comment by tylerjereddy** [comment]
@peterjc123 I've opened a debug PR to pytorch to see if our latest Windows Python 3.6 wheel, containing all the attempted fixes, will pass on your circleci run: https://github.com/pytorch/pytorch/pull/40975

I hope that doesn't cause too much trouble for pytorch team--it is one of the more obvious ways for me to check the fix (I tried setting up CI on a fork of pytorch, but this seemed complex).

**Comment by peterjc123** [comment]
> @peterjc123 I've opened a debug PR to pytorch to see if our latest Windows Python 3.6 wheel, containing all the attempted fixes, will pass on your circleci run: [pytorch/pytorch#40975](https://github.com/pytorch/pytorch/pull/40975)
> 
> I hope that doesn't cause too much trouble for pytorch team--it is one of the more obvious ways for me to check the fix (I tried setting up CI on a fork of pytorch, but this seemed complex).

Thanks. I'll handle this.

**Comment by tylerjereddy** [comment]
The Python 3.6 Windows DLL load issue should be fixed in the recently-released SciPy `1.5.1`, based in part on the testing in pytorch CI cross-linked above.

I'll close this tentatively, please feel free to open another issue if DLL handling still has problems in `1.5.1`.
